README.md
# vft-data-sanitizer

Removes sensitive information from XML files containing visual field test data 
and replaces patient names by number ids using fuzzy matching.


# Installation

```
git clone https://github.com/maet3608/vft-data-sanitizer.git
cd vft-data-sanitizer
python setup.py develop
```


# Usage

The software will recursively scan the ``data_folder`` for files ending with
``_SFA.xml``, will remove all sensitive patient data in these files,
will replace the patient name by a unique ID (provided in the file 
``csv_file_with_subject_id``) and writes the sanitized files to the
``output_folder``.

```bash
cd vft-data-sanitizer/vds
python sanitize.py <csv_file_with_subject_ids> <data_folder> <output_folder>
```


# Example

Run sanitize.py on example data.
 
```bash
cd vft-data-sanitizer/vds
python sanitize.py ../data/index.csv ../data ../out
```

Output should look like this

```bash
running sanitizer version 2.2.1
loading index ../data/index.csv
processing 2 files ...
1 of 2 : DOE_20121024_114922_OD_000000_SFA.xml -> 000011-2012-10-24-11-49-22-OD.xml
2 of 2 : DOE_20131118_092712_OD_00001_SFA.xml -> 000007-2013-11-18-09-27-12-OD.xml
finished with 0 error(s)
```

If there are any errors, please check the logfile `sanitize.log`




RELEASENOTES.md
# Release Notes


### Version 2.2.1

- fixed: detection of duplicates in output


### Version 2.2.0

- recursively scan data folder for XML files.


### Version 2.1.0

- replace fullname by subject id in XML file


### Version 2.0.0

- map patient names to subject ids 
- sanitize filenames


### Version 1.0.0

- first version. Removes sensitive data


vds/fuzzy.py
"""
Fuzzy matching of strings/names.
"""

import csv
from collections import defaultdict, Counter


def group_by(kv_pairs):
    """Group key-value pairs by key"""
    groups = defaultdict(list)
    for k, v in kv_pairs:
        groups[k].append(v)
    return groups


def ngrams(text, n=3):
    """Return list of text n-grams of size n"""
    return {text[i:i + n].lower() for i in range(len(text) - n + 1)}


def find(query, index):
    """Find best match of query string in n-gram index"""
    sgrams = ngrams(query)
    matches = Counter(sid for ngram in sgrams for sid in index.get(ngram, []))
    return matches.most_common(1)


def create_index(filepath):
    """Create n-gram index that maps n-grams to subject ids"""
    sid2name = load_sid2name(filepath)
    return group_by((ngram, sid) for sid, name in sid2name for ngram in ngrams(name))


def load_sid2name(filepath):
    """Load file that maps subject ids to names"""
    with open(filepath) as csvfile:
        reader = csv.reader(csvfile)
        next(reader)  # skip header
        for sid, name in reader:
            yield sid.strip(), name.replace(',', ' ')


# Example main
if __name__ == '__main__':
    print('running')
    index = create_index('../data/index.csv')
    while 1:
        query = input('search: ')
        if query == 'q': break
        match = find(query, index)
        print(match[0] if match else 'NO MATCH')
        print('-' * 50)


vds/sanitize.py
"""
Removes tags with patient info from XML file and replaces patient names by
subject ids uses fuzzy matching between patient names in XML file and
the index database.
"""

import logging
import datetime
import fuzzy
import vds

import os, sys
import os.path as osp

logging.basicConfig(filename='sanitize.log',
                    filemode='w',
                    format='%(levelname)s : %(message)s',
                    level=logging.INFO)
logger = logging.getLogger('sanitizer')

error_counter = 0  # number file conversion errors
outpaths = set()  # Keep track of output files created

# Full name tags
SNTAG, ENTAG = '<FULL_NAME>', '</FULL_NAME>'

# Set of tags to remove. Note that full name will be replaced by sid
STAGS = {'<LAST_NAME', '<GIVEN_NAME', '<MIDDLE_NAME', '<NAME_PREFIX',
         '<NAME_SUFFIX', '<BIRTH_DATE', '<PATIENT_ID',
         '<IMAGE_FILE_NAME'}


def log_info(msg):
    """Log and print info messages"""
    print(msg)
    logger.info(msg)


def extract_patientname(filepath):
    """Extract full name from xml file"""
    with open(filepath) as f:
        for line in f:
            if line.startswith(SNTAG):
                name = line.strip().replace(SNTAG, '').replace(ENTAG, '')
                return name
    raise ValueError('No full patient name in ' + filepath)


def split_filename(filename):
    """Splits filenames of form DOE_20121024_114922_OD_000000_SFA.xml
    into their components and checks for validity"""
    name, ext = filename.split('.')
    try:
        sname, vdate, vtime, lat, sid, kind = name.split('_')
    except:
        raise ValueError('Invalid filename :' + filename)
    if ext != 'xml':
        raise ValueError('Not an .xml file :' + filename)
    if kind != 'SFA':
        raise ValueError('Not a SFA file :' + filename)
    if lat not in ['OD', 'OS']:
        raise ValueError('Unknown eye laterality :' + filename)
    if len(vdate) != 8:
        raise ValueError('Invalid visit date :' + filename)
    if len(vtime) != 6:
        raise ValueError('Invalid visit time :' + filename)
    return vdate, vtime, lat


def create_filename(sid, vdate, vtime, lat):
    """Creates an output filename with long uid based on components of
    input filename, e.g.
    DOE_20121024_114922_OD_000000_SFA.xml -> 000000-2012-10-24-11-49-22-OD.xml
    """
    y, o, d = vdate[:4], vdate[4:6], vdate[6:8]
    h, m, s = vtime[:2], vtime[2:4], vtime[4:6],
    uid = '%s-%s-%s-%s-%s-%s-%s' % (y, o, d, h, m, s, lat)
    filename = '%s-%s.xml' % (sid, uid)
    return filename


def is_sensitive(line):
    """Return true if line (in XML filed) starts with a sensitive TAG"""
    return any(line.startswith(t) for t in STAGS)


def remove_sensitive(lines):
    """Return iterator over lines skipping lines with sensitive data"""
    return (l for l in lines if not is_sensitive(l))


def sanitize_file(infilepath, outfilepath, sid):
    """Copy sanitized from infilepath to outfilepath"""
    with open(outfilepath, 'w') as fout:
        lines = open(infilepath)
        for line in remove_sensitive(lines):
            if line.startswith(SNTAG):  # fullname found
                line = '%s%s%s\n' % (SNTAG, sid, ENTAG)
            fout.write(line)


def name2sid(index, name, filename):
    """Map name to subject id using fuzzy matching"""
    match = fuzzy.find(name, index)
    logging.info('matches for %s : %s' % (name, str(match)))
    if match and match[0][1] > 1:
        return '%06d' % int(match[0][0])
    raise ValueError("Could not find sid for '%s' in %s" % (name, filename))


def sanitize(index, infilepath, outdir):
    """Remove sensitiv data and map name to id"""
    try:
        infilename = osp.basename(infilepath)
        name = extract_patientname(infilepath)
        sid = name2sid(index, name, infilename)
        vdate, vtime, lat = split_filename(infilename)

        outfilename = create_filename(sid, vdate, vtime, lat)
        outfilepath = osp.join(outdir, outfilename)
        if outfilepath in outpaths:
            raise ValueError('Duplicate output %s ' % outfilepath)
        outpaths.add(outfilepath)

        logging.info('writing %s' % outfilename)
        sanitize_file(infilepath, outfilepath, sid)
        return outfilename
    except Exception as e:
        global error_counter
        error_counter += 1
        logger.error(str(e))
        print(e)
        return str(e)


def xmlfile_check(indir, outdir):
    """Return list of XML filepaths and check data/dir existence"""
    infilepaths = []
    for path, dirs, files in os.walk(indir):
        fpaths = [osp.join(path, f) for f in files if f.endswith('_SFA.xml')]
        infilepaths.extend(fpaths)

    if not infilepaths:
        raise IOError('No .xml files in: ' + indir)
    if not osp.isdir(outdir):
        raise IOError('Output dir does not exist: ' + outdir)

    return infilepaths


def run(indexfile, indir, outdir):
    """Sanitized all XML files in indir and copy to outdir"""
    log_info('loading index ' + indexfile)
    index = fuzzy.create_index(indexfile)

    infilepaths = xmlfile_check(indir, outdir)
    n = len(infilepaths)
    log_info('processing %d files ...' % n)
    for i, infilepath in enumerate(infilepaths):
        logger.info('processing ' + infilepath)
        infilename = osp.basename(infilepath)
        outfilename = sanitize(index, infilepath, outdir)
        print('%d of %d : %s -> %s' % (i + 1, n, infilename, outfilename))

    log_info('finished with %d error(s)' % error_counter)


if __name__ == '__main__':
    log_info("running sanitizer version %s" % vds.__version__)

    now = datetime.datetime.now()
    logger.info('time of processing ' + now.strftime("%Y-%m-%d %H:%M"))

    try:
        logger.info('cmdline args: ' + ' '.join(sys.argv))
        if len(sys.argv) != 4:
            raise IOError('Expect index file, input and output folder')
        _, indexfile, indir, outdir = sys.argv
        run(indexfile, indir, outdir)
    except Exception as e:
        logger.error(str(e))
        print(e)
